OpenAL 32-bit fixed point support

LAGonauta opened the pull request:
I believe that clamping to (-1 ; 1) is the best solution, and this new method works correctly, even though it is possible to hear the hard-clipping it is better than distortion. When the DPL2 decoder is reworked the clamping will probably be unnecessary, but until then it is needed.

Original PR message follows:

The X-Fi DSP supports only 32-bit and 16-bit fixed point buffers. This commit adds support to 32-bit fixed point buffers on stereo and on surround and allows a X-Fi to output surround sound with no artifacts or noise. I also changed how floating point support is detected.
There is a problem, though. The Dolby Decoder gives some samples outside the usual [ -1 ; 1 ] range, and just clamping them to 1 or -1 gives undesirable sound artifacts as the samples are usually at [ -2,5 ; 2,5 ]. As there is enough dynamic range on 32-bit fixed point, I am clamping it to [ -4 ; 4 ]. The best solution is to fix the DPL2 decoder somehow, but I am still trying to understand how it works.
FFDShow always clamps to one, but I did not verify if its DPL2 decoder gives samples outside [ -1; 1 ].

Tested on OpenAL Soft, X-Fi and Generic Software (wrap_oal). Everything seems to be working, but someone needs to test it on Linux and on Mac.

EDIT:
I am not clamping on [ -4; 4 ] anymore as this is clearly wrong, but this causes some artifacts. I need to fix the DPL2 decoder somehow.

<!-- Reviewable:start -->

[<img src="https://reviewable.io/review_button.svg" height="40" alt="Review on Reviewable"/>](https://reviewable.io/reviews/dolphin-emu/dolphin/3080)

<!-- Reviewable:end -->

LAGonauta commented:
I wonder why it fails to render... Any ideas?

phire commented:
That was before I fixed the cause of most of our random fifoci failures.

Should be safe to ignore that.
On Sep 30, 2015 4:13 PM, "LAGonauta" <notifications@github.com> wrote:

> I wonder why it fails to render... Any ideas?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/dolphin-emu/dolphin/pull/3080#issuecomment-144270057>.
>

skidau commented on Source/Core/AudioCommon/OpenALStream.cpp:
```diff
@@ -172,12 +191,19 @@ void OpenALStream::SoundLoop()
-			alBufferData(uiBuffers[i], AL_FORMAT_STEREO16, realtimeBuffer, 4 * FRAME_STEREO_SHORT, ulFrequency);
+			if (float32_capable)
+				alBufferData(uiBuffers[i], AL_FORMAT_STEREO_FLOAT32, realtimeBuffer, 4 * FRAME_STEREO_FLOAT, ulFrequency);
+			if (fixed32_capable)
```
Should this line be "else if (fixed32_capable)"  to match the ones above?

skidau commented on Source/Core/AudioCommon/OpenALStream.cpp:
```diff
@@ -132,15 +132,21 @@ void OpenALStream::SoundLoop()
+#define AL_FORMAT_STEREO32 0x1203
 #else
-	bool float32_capable = true;
+	const ALenum AL_FORMAT_STEREO32 = 0
```
Missing a semi-colon ; at the end of this line

bb010g commented:
Any chance of getting this done soon? The higher quality would be nice to have for 5.0.

JosJuice commented:
@dolphin-emu-bot rebuild

LAGonauta commented on Source/Core/AudioCommon/OpenALStream.cpp:
```diff
@@ -285,16 +311,41 @@ void OpenALStream::SoundLoop()
+						else if (dpl2[i] <= -1)
+							surround_int32[i] = INT_MIN;
+						else
+							surround_int32[i] = (int)((float)dpl2[i] * (1 << 31));
```
Sadly I will not able to test again with my X-Fi until 26/03, but if I remember one problem I was having when clamping to [-1,1] is that when the value is close enough to 1 or -1 (like 0.9999999999999 and -0.9999999999999) the operation would do something wrong and cause all sorts of clicks. Anyone knows why? If I multiply by 2147483648.0 (a double) instead of using bit-shifting this problem does not happen.

degasus replied:
1 << 31 is a negative number. Maybe 1LL ?

lioncash replied:
`INT64_C(1)` is the standardized non-cast way to do it.

LAGonauta replied:
Thanks, that makes sense and should work :)
If someone with a X-Fi can test it please do so. 
I will amend the commit again with 1LL.

degasus commented on Source/Core/AudioCommon/OpenALStream.cpp:
```diff
@@ -285,16 +311,41 @@ void OpenALStream::SoundLoop()
+						else if (dpl2[i] <= -1)
+							surround_short[i] = SHRT_MIN;
+						else
+							surround_short[i] = (int)((float)dpl2[i] * (1 << 15));
```
Here is another kind of issue. There are values too close to 1.0, but still smaller 1.0. For example 1.0 - 1.0 / (1<<23) can be represented as a float. So you'll end with 1.0 * (1<<15), which results in -(1<<15) when casted to short.

degasus commented on Source/Core/AudioCommon/OpenALStream.cpp:
```diff
@@ -322,12 +373,25 @@ void OpenALStream::SoundLoop()
+					for (u32 i = 0; i < nSamples * STEREO_CHANNELS; ++i)
+					{
+						// Clamping is not necessary here, samples are always between (-1,1)
+						stereo_int32[i] = (int)((float)sampleBuffer[i] * (1 << 31));
```
same here, integer overflow.

LAGonauta replied:
By my tests sampleBuffer never got close enough to 1.0 for that to be a problem, and I assumed it would work as this is how it was originally written on line 395 (but for 16-bit fixed point). Of course it should have probably been something like (1LL << 31) or (INT64_C(1) << 31).

RisingFog commented:
What's the status of this PR?

LAGonauta commented:
I need to rebase and resolve all conflicts to merge, but it works perfectly for stereo output and clips for surround as the DPL2 decoder outputs samples outside the ]-1,1[ range.
I also do not like to use `(INT64_C(1) << 31) - 1)` on line 326 and `((1 << 15) - 1))` on line 342, but this is the only way to convert back to integer without artifacts that probably come from the truncation.
I can change the code to something like this:

```c++
float dpl2_toint;
int surround_int32[OAL_MAX_SAMPLES * SURROUND_CHANNELS * OAL_MAX_BUFFERS];

for (u32 i = 0; i < nSamples * SURROUND_CHANNELS; ++i)
{
  dpl2_toint = dpl2[i] * (INT64_C(1) << 31);

  if (dpl2_toint > INT_MAX)
    surround_int32[i] = INT_MAX;
  else if (dpl2_toint < INT_MIN)
    surround_int32[i] = INT_MIN;
  else
    surround_int32[i] = (int)dpl2_toint[i];
}
```

This works fine, but I am not sure if it is the best solution.

LAGonauta commented:
I just rebased this PR and added some changes to how I convert 32-bit floating point data to 32-bit fixed point data.
I believe this is good enough until the DPL2 decoder is reworked.

shuffle2 commented:
@dolphin-emu-bot rebuild

shuffle2 commented:
@dolphin-emu-bot rebuild

shuffle2 commented on Source/Core/AudioCommon/OpenALStream.cpp:
```diff
@@ -149,10 +155,13 @@ void OpenALStream::SoundLoop()
+  // is the only one that supports it for now.
   if (strstr(alGetString(AL_RENDERER), "X-Fi"))
-    float32_capable = false;
+    fixed32_capable = true;
```
You have removed the code which set float32_capable to true. Is it dead now?

LAGonauta replied:
Actually, the X-Fi driver never reported an error when buffering 32-bit floating point data (OpenAL was never able to fall back to 16-bit fixed point buffers), so it played this data as 32-bit fixed point creating disgusting noises. 
I was doing it wrong before, I should have checked for the extensions instead of forcing any X-Fi to not use 32-bit floating point buffers:
```
if (!alIsExtensionPresent("AL_EXT_float32"))
    float32_capable = false;
```

But now I prefer having the 32-bit floating point support normally off, and just checking if the OpenAL driver has this extension to enable support.
The X-Fi driver never supported 32-bit floating point data, but has undocumented support for 32-bit fixed point. There is no extension to check for support, but I believe all X-Fi with hardware OpenAL supports this feature.

"OpenAL Soft" supports `AL_EXT_float32` so it works fine, "Generic Software" doesn't so it falls back to 16-bit fixed point.

mirh replied:
Harcoding for vendor seems quite dengerous if you ask me.. 
Can anybody confirm this still applies to nvidia nforce implementation, cmedia, asus and realtek ones?

LAGonauta replied:
I believe the nForce implementation is quite dead since... 2004? I bet it doesn't even work on Vista or newer. Forget about any 64-bit processor support.

ASUS seems to use the implementation from C-Media, I hope someone can test it but I bet that it only supports 16-bit fixed point, with luck it also supports floating point. It is probably software rendered and properly fall back to 16-bit if floating point support is not found. It would be great if someone could test it. Also, does it have 64-bit processor support?

My onboard audio is from Realtek, and no OpenAL driver comes with the newest driver. Is it still used/supported by Realtek?

Some Creative cards use software rendered OpenAL from Sensaura, through Sens_oal.dll driver. Last time I tested it didn't support any kind of 32-bit samples, but works fine with the 16-bit fall back.

I believe that valid software rendered drivers that should be tested are "OpenAL Soft", "AeonWave-OpenAL", "Rapture3D" and the "Generic software" from default wrap_oal.dll driver.
OpenAL Soft works great with floating point samples as anyone can test it and "Generic software" from wrap_oal.dll properly falls back to 16-bit, but it would be great if someone could test with those other renderers.

LAGonauta replied:
Yes, you are right. Anything after X-Fi (and some cards branded as X-Fi) from Creative uses only Sensaura's OpenAL driver.

Just tried 3DSoundBack. It only has a 32-bit OpenAL driver so it won't work with Dolphin.

Sadly RightMark is not enough, but you can at least test if there is any 64-bit support by going to your system32 folder and looking for a "cm_oal.dll" or any other file ending with "_oal.dll" which is not "soft_oal.dll" and "wrap_oal.dll" (if you are on a 64-bit Windows).

mirh replied:
Oh right, I had forgot Dolphin is 64 bit (never liked the [explanation](https://it.dolphin-emu.org/blog/2014/05/19/obituary-32bit/) for this tbh) :s 

I wonder if with respect to OpenAL soft there's any actual downside then.

BhaaLseN replied:
Can you extend the comment to explain why you (have to) check for the name instead of a registered extension or anything like that?

shuffle2 commented on Source/Core/AudioCommon/OpenALStream.cpp:
```diff
@@ -132,15 +132,21 @@ void OpenALStream::SoundLoop()
   const ALenum AL_FORMAT_51CHN16 = 0;
+#elif defined(_WIN32)
+  // Only Windows supports AL_FORMAT_STEREO32 alext for now.
+  const ALenum AL_FORMAT_STEREO32 = 0x1203;
```
Having varying definitions of this "constant" is confusing. Instead you should probably have a constant variable like `m_format` or whatever, and assign the proper OAL constant to it...

LAGonauta replied:
Sorry, I did not understand your request correctly. With `m_format` you mean member variable?
Maybe it is better to move this:
```
#if defined(__APPLE__)
  const ALenum AL_FORMAT_STEREO_FLOAT32 = 0;
  const ALenum AL_FORMAT_STEREO32 = 0;
  // OS X does not have the alext AL_FORMAT_51CHN32 yet.
  surround_capable = false;
  const ALenum AL_FORMAT_51CHN32 = 0;
  const ALenum AL_FORMAT_51CHN16 = 0;
#elif defined(_WIN32)
  // Only Windows supports AL_FORMAT_STEREO32 alext for now.
  const ALenum AL_FORMAT_STEREO32 = 0x1203;
#else
  const ALenum AL_FORMAT_STEREO32 = 0;
#endif
```
To the header file? And change it to something like this:
```
#if defined(__APPLE__)
// OS X does not have the alext AL_FORMAT_STEREO_FLOAT32, AL_FORMAT_STEREO32, AL_FORMAT_51CHN32 and AL_FORMAT_51CHN16 yet.
#define AL_FORMAT_STEREO_FLOAT32 = 0;
#define AL_FORMAT_STEREO32 = 0;
#define AL_FORMAT_51CHN32 = 0;
#define AL_FORMAT_51CHN16 = 0;
#elif defined(_WIN32)
// Only Windows supports AL_FORMAT_STEREO32 alext for now.
#define AL_FORMAT_STEREO32 = 0x1203;
#else
#define AL_FORMAT_STEREO32 = 0;
#endif
```

But keep this on OpenALStream.cpp:
```
#if defined(__APPLE__)
  surround_capable = false;
```

LAGonauta commented:
Review status: 0 of 2 files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[Source/Core/AudioCommon/OpenALStream.cpp, line 149 at r1](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3080#-KBaaTUs9tkYQyCK-JBR-r1-149:-KTtONs-uClqw3DzTcFu:b-896fix) ([raw file](https://github.com/dolphin-emu/dolphin/blob/f3245abdbac6eea4ecf1384d2f823c12a888ea3c/Source/Core/AudioCommon/OpenALStream.cpp#L149)):*
<details><summary><i>Previously, skidau wrote…</i></summary>
> Missing a semi-colon ; at the end of this line
</details>
Done.

---

*[Source/Core/AudioCommon/OpenALStream.cpp, line 326 at r2](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3080#-KBaaTUs9tkYQyCK-JBR-r2-326:-KTHaMnElBwo-XZE8-d-:b-896fix) ([raw file](https://github.com/dolphin-emu/dolphin/blob/5e6d3cbff878e1cdbcab348cabaaae480c20ac35/Source/Core/AudioCommon/OpenALStream.cpp#L326)):*
<details><summary><i>Previously, LAGonauta wrote…</i></summary>
> Thanks, that makes sense and should work :)
> 
> If someone with a X-Fi can test it please do so. 
> 
> I will amend the commit again with INT64_C(1).
</details>
Done.

---

*[Source/Core/AudioCommon/OpenALStream.cpp, line 147 at r3](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3080#-KBaaTUs9tkYQyCK-JBR-r3-147:-KV-lziKUCDN-NAFRO1S:bs6ny2z) ([raw file](https://github.com/dolphin-emu/dolphin/blob/68e13c47bab82220ae57b4339aeb99a0e4526d4c/Source/Core/AudioCommon/OpenALStream.cpp#L147)):*
<details><summary><i>Previously, LAGonauta wrote…</i></summary>
> Sorry, I did not understand your request correctly. With `m_format` you mean member variable?
> 
> Maybe it is better to move this:
> 
> ```
> 
> #if defined(__APPLE__)
> 
>   const ALenum AL_FORMAT_STEREO_FLOAT32 = 0;
> 
>   const ALenum AL_FORMAT_STEREO32 = 0;
> 
>   // OS X does not have the alext AL_FORMAT_51CHN32 yet.
> 
>   surround_capable = false;
> 
>   const ALenum AL_FORMAT_51CHN32 = 0;
> 
>   const ALenum AL_FORMAT_51CHN16 = 0;
> 
> #elif defined(_WIN32)
> 
>   // Only Windows supports AL_FORMAT_STEREO32 alext for now.
> 
>   const ALenum AL_FORMAT_STEREO32 = 0x1203;
> 
> #else
> 
>   const ALenum AL_FORMAT_STEREO32 = 0;
> 
> #endif
> 
> ```
> 
> To the header file? And change it to something like this:
> 
> ```
> 
> #if defined(__APPLE__)
> 
> // OS X does not have the alext AL_FORMAT_STEREO_FLOAT32, AL_FORMAT_STEREO32, AL_FORMAT_51CHN32 and AL_FORMAT_51CHN16 yet.
> 
> #define AL_FORMAT_STEREO_FLOAT32 = 0;
> 
> #define AL_FORMAT_STEREO32 = 0;
> 
> #define AL_FORMAT_51CHN32 = 0;
> 
> #define AL_FORMAT_51CHN16 = 0;
> 
> #elif defined(_WIN32)
> 
> // Only Windows supports AL_FORMAT_STEREO32 alext for now.
> 
> #define AL_FORMAT_STEREO32 = 0x1203;
> 
> #else
> 
> #define AL_FORMAT_STEREO32 = 0;
> 
> #endif
> 
> ```
> 
> 
> 
> But keep this on OpenALStream.cpp:
> 
> ```
> 
> #if defined(__APPLE__)
> 
>   surround_capable = false;
> 
> ```
</details>
@shuffle2 Is it better now? It compiled on Windows, I am not sure if it will compile on Mac or Linux.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3080)*
<!-- Sent from Reviewable.io -->

JosJuice commented:
@dolphin-emu-bot rebuild

LAGonauta commented:
Review status: 0 of 2 files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[Source/Core/AudioCommon/OpenALStream.cpp, line 164 at r3](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3080#-KBaaTUs9tkYQyCK-JBR-r3-164:-KV6to1_tqLKuTOSfcKQ:b-896fix) ([raw file](https://github.com/dolphin-emu/dolphin/blob/68e13c47bab82220ae57b4339aeb99a0e4526d4c/Source/Core/AudioCommon/OpenALStream.cpp#L164)):*
<details><summary><i>Previously, BhaaLseN (BhaaL) wrote…</i></summary>
> Can you extend the comment to explain why you (have to) check for the name instead of a registered extension or anything like that?
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3080)*
<!-- Sent from Reviewable.io -->

shuffle2 commented:
@dolphin-emu-bot rebuild
